src/base.jl

```diff
@@ -280,15 +284,116 @@ function add_component!(sys::ConcreteSystem, component::T) where T <: Component
+                num_not_unique += 1
+                match = false
+                @error("not all components in forecast vector are unique", component_label,
+                       issue_time)
+            else
+                push!(unique_components, component_label)
+            end
+        end
+        @info "summary" num_not_unique length(forecasts)
+    end
+
+    return match
+end
+
+function add_forecast!(forecasts::SystemForecasts, issue_time::IssueTime,
+                       forecast::Forecast)
+    if !haskey(forecasts, issue_time)
+        forecasts[issue_time] = Forecasts{Forecast}()
+    end
+
+    push!(forecasts[issue_time], forecast)
+end
+
+function add_forecast!(forecasts::SystemForecasts, forecast::Forecast)
+    issue_time = get_issue_time(forecast)
+    add_forecast!(forecasts, issue_time, forecast)
+end
+
 """
 Args:
     A ConcreteSystem struct
     A :Symbol=>Array{ <: Forecast,1} Pair denoting the forecast name and array of device forecasts
 Returns:
     A System struct with a modified forecasts field
 """
-function add_forecast!(sys::ConcreteSystem,fc::Pair{Symbol,Array{Forecast,1}})
-    sys.forecasts[fc.first] = fc.second
+function add_forecast!(sys::ConcreteSystem, fc::Pair{Symbol, Vector{Forecast}})
+    tmp = SystemForecastsDeprecated()
+    tmp[fc.first] = fc.second
+    system_forecasts = SystemForecasts()
+    _convert_forecasts!(system_forecasts, tmp)
+
+    for (issue_time, forecasts) in system_forecasts
+        # This assumes that issue_time is the same for each forecast in the vector because
+        # that is how _convert_forecasts works above.
+        for forecast in forecasts
+            add_forecast!(sys.forecasts, issue_time, forecast)
+        end
+    end
+
+    if !_validate_component_label_uniqueness(sys.forecasts)
+        throw(DataFormatError("components and labels are not unique within forecast array"))
+    end
+end
+
+"""Returns an iterator to the forecast IssueTime values stored in the System."""
+function get_forecast_issue_times(sys::ConcreteSystem)
+    return keys(sys.forecasts)
+end
+
+"""Returns an iterator to the forecasts for the given IssueTime stored in the System."""
+function get_forecasts(sys::ConcreteSystem, issue_time::IssueTime)
+    if !haskey(sys.forecasts, issue_time)
+        error("forecast issue_time {issue_time} does not exist")
+    end
+
+    return Iterators.take(sys.forecasts[issue_time], length(sys.forecasts[issue_time]))
+end
+
+"""Returns a vector of forecasts that match the components and label."""
+function get_forecasts(sys::ConcreteSystem, issue_time::IssueTime, components_iterator,
+                       label::Union{String, Nothing}=nothing)
+    forecasts = Vector{Forecast}()
+
+    # This code caches the passed component UUIDs in a dict so that we don't have to
+    # iterate across them for each forecast.
+    components = LazyDictFromIterator(Base.UUID, Component, components_iterator, get_uuid)
```

Comment:
Another option is to let the user create the LazyDictFromIterator and pass it in.  This could be useful if they are going to call this function multiple times with the same components.  However, I don't like the added complexity.  I think LazyDictFromIterator should remain an implementation detail rather than a public object.